cmd/clusterctl/cmd/topology.go

```diff
@@ -0,0 +1,306 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"sort"
+
+	"github.com/olekukonko/tablewriter"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	utilyaml "sigs.k8s.io/cluster-api/util/yaml"
+)
+
+type topologyDryRunOptions struct {
+	file    string
+	cluster string
+	outDir  string
+}
+
+var dr = &topologyDryRunOptions{}
+
+var topologyDryRunCmd = &cobra.Command{
+	Use:   "topology-dryrun",
+	Short: "Dry run changes to clusters that use managed topologies.",
+	Long: LongDesc(`
+	    Provide the list of objects that would be created, modified and deleted when a change is applied.
+		The objects that will be created and modified will be stored in a path passed using --outdir.
+	`),
+	Example: Examples(`
+	    # List the changes when modifying a cluster.
+		clusterctl alpha topology-dryrun -f modified-cluster.yaml --outdir output/
+
+		# List the changes of a single cluster when multiple clusters are affected.
+		clusterctl alpha topology-dryrun -f modified-cluster-class.yaml --cluster "cluster1" --outdir output/
+	`),
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runTopologyDryRun()
+	},
+}
+
+func init() {
+	topologyDryRunCmd.Flags().StringVarP(&dr.file, "file", "f", "", "path to the input file")
+	topologyDryRunCmd.Flags().StringVarP(&dr.cluster, "cluster", "c", "", "name of the target cluster")
+	topologyDryRunCmd.Flags().StringVar(&dr.outDir, "outdir", "", "output directory to write created/modified objects")
+
+	if err := topologyDryRunCmd.MarkFlagRequired("outdir"); err != nil {
+		panic(err)
+	}
+
+	alphaCmd.AddCommand(topologyDryRunCmd)
+}
+
+func runTopologyDryRun() error {
+	c, err := client.New(cfgFile)
+	if err != nil {
+		return err
+	}
+
+	raw, err := os.ReadFile(dr.file)
+	if err != nil {
+		return errors.Wrap(err, "failed to read input file")
+	}
+
+	objs, err := utilyaml.ToUnstructured(raw)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert file to list of objects")
+	}
+
+	out, err := c.DryRunTopology(client.DryRunOptions{
+		Objs:    convertToPtrSlice(objs),
+		Cluster: dr.cluster,
+	})
+	if err != nil {
+		return err
+	}
+	return handleDryrunOutput(out, dr.outDir)
```

Comment:
I originally had `printDryrunOutput` but renamed it to `handleDryrunOutput` because we also do things like write files and not exclusively print. I dont have string opinions and can go back to `printDryRunOutput`.